Handle more than 8 arguments in CheriABI

kwitaszczyk opened the pull request:
Function calls with more than 8 arguments require passing 8th and later arguments through memory referenced by the c9 register.

libffi has already passed such arguments on the stack. Extend the implementation to store a pointer to the stack in c9 with bounds restricted to the in-memory arguments.

Discussed with:	@dpgao

kwitaszczyk committed:
Handle more than 8 arguments in CheriABI

Function calls with more than 8 arguments require passing 9th and later
arguments through memory referenced by the c9 register.

libffi has already passed such arguments on the stack. Extend the
implementation to store a pointer to the stack in c9 with bounds
restricted to the in-memory arguments. Additionally, use precise bounds
for all objects passed to ffi_call_SYSV(). They are allocated together
with the on-stack arguments and we want to make sure they do not overlap
with anything that should not be leaked to the callee.

Before we allocate memory for the on-stack arguments, we must first
compute the layout of data to be stored in memory that would meet
requirements of precise bounds to be used with that data. The number of
requested stack bytes is calculcated prior to ffi_call_int() and is
sufficient to hold on-stack function arguments and structure objects
passed by value. libffi allocates memory for a call context, on-stack
arguments, a call frame and return values on the caller's stack using
alloca(). In order to align the on-stack arguments to use precise
bounds, all sub-objects within that allocation must be appropriately
padded with respect to other sub-objects' alignments.  Otherwise, if
only the on-stack arguments were aligned, sub-objects placed after the
on-stack arguments could be misaligned.

After that, the stack can be further narrowed to use precise bounds only
for on-stack function arguments, excluding the structure objects passed
by value. Since the capability alignment grows with length and is
divisible by smaller alignments, it is safe to do that.

Co-authored-by:	Dapeng Gao <dapeng.gao@cl.cam.ac.uk>
Co-authored-by: Jessica Clarke <jessica.clarke@cl.cam.ac.uk>

arichardson reviewed (commented):
Do we actually need to store this in the context? Couldn't we just pass an additional argument to the assembly function?

dpgao commented:
> Do we actually need to store this in the context? Couldn't we just pass an additional argument to the assembly function?

We considered doing that, but it seemed less clean than just adding a field to the context.

kwitaszczyk commented:
> Do we actually need to store this in the context? Couldn't we just pass an additional argument to the assembly function?

I agree it would be simpler but it would make libffi's API inconsistent between ABIs. Do you think it's safe to do that and we can treat `ffi_call_SYSV` as an internal interface that can be modified for each ABI?

kwitaszczyk commented:
I can see `ffi_call_SYSV` is declared with `FFI_HIDDEN` so it might be safe to change this interface.

dpgao commented:
@kwitaszczyk and I discussed. It seemed to us that adding a field to the context is the consistent thing to do here because `c9` is really an argument in this ABI, so it belongs to the struct holding arguments.

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -927,6 +930,10 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
     }
 
+#ifdef __CHERI_PURE_CAPABILITY__
+  context->x9 = (XREG)__builtin_cheri_bounds_set(stack, state.nsaa);
```
This doesn't handle the case where it's not representable. In that instance, the stack region (a) needs to be overaligned (b) needs to be padded.

kwitaszczyk commented:
I'm converting this to a draft as I must first test it with the GUI stack on a Morello box that I don't have right now.

dpgao commented on src/aarch64/ffi.c:
```diff
@@ -731,9 +855,19 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
     rsize = 16;
 
+#ifdef __CHERI_PURE_CAPABILITY__
+  stack_argsize = arg_size(cif, stack_bytes, isvariadic);
+  stack_padding = __builtin_cheri_round_representable_length(stack_argsize) -
+    stack_argsize;
```
Defer the computation of `stack_argsize` and just do the following
```
stack_padding = __builtin_cheri_round_representable_length(stack_bytes) - stack_bytes;
```

kwitaszczyk replied:
The current approach does this.

dpgao commented on src/aarch64/ffi.c:
```diff
@@ -743,6 +877,13 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
    * points just after the alloca with the curretn stack bounds.
    */
   stack = __builtin_cheri_address_set(__builtin_cheri_stack_get(), (ptraddr_t)stack);
+  /*
+   * Align the stack so that the c9 capability to on-stack function arguments
+   * can be bounded exactly to the number of bytes occupied by them.
+   */
+  stack = __align_up(stack,
+		     ~__builtin_cheri_representable_alignment_mask(stack_argsize) +
+		     1);
```
Defer this step till after the memory arguments have been allocated.

kwitaszczyk replied:
We need `stack` before we store arguments and it must point at the same location as it will after being aligned.

dpgao reviewed (commented):
I have an idea which might save the duplication of code. Essentially, we treat `stack_bytes` as an over-estimation of the memory-argument buffer and insert padding accordingly.

Then, when allocating arguments to the memory-argument buffer, we compute the precise `stack_argsize`.

Finally, we bound `stack` to `stack_argsize`.

kwitaszczyk commented:
> I have an idea which might save the duplication of code. Essentially, we treat `stack_bytes` as an over-estimation of the memory-argument buffer and insert padding accordingly.
> 
> Then, when allocating arguments to the memory-argument buffer, we compute the precise `stack_argsize`.
> 
> Finally, we bound `stack` to `stack_argsize`.

My initial thought was that `stack_bytes` is carefully calculated to include other things than just on-stack arguments and on-stack structure objects passed by value. After reading `ffi_prep_cif_core()` that is used to calculate `stack_bytes` prior to a `ffi_call_int()` call, I can see the on-stack buffer is only used for these values.

I've changed the implementation to pad the stack according to the original `stack_bytes` length but later narrow the bounds to the effective length.

This change is tested with Xwayland now.

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -728,9 +734,18 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
-  context = alloca (sizeof(struct call_context) + stack_bytes + sizeof(struct call_frame) + rsize);
+  context = alloca (sizeof(struct call_context) +
+#ifdef __CHERI_PURE_CAPABILITY__
+		    stack_padding +
```
Just because the length is representable doesn't mean the bounds will be at an offset of `sizeof(struct call_context)` within the allocation (only guaranteed that the whole allocation is).

jrtc27 replied:
(In an extreme, consider sizeof(struct call_context) is 1, stack_bytes is 2^32, then there's no padding but your requested base for the subregion will have an LSB of 1, which isn't representable for a length of 2^32)

dpgao replied:
I think that further adding `~__builtin_cheri_representable_alignment_mask(stack_bytes)` to `stack_padding` would be a simple fix to this, albeit wasting some space.

dpgao replied:
Oh, but then this would cause the `struct call_frame` to be completely misaligned. We should probably add `~__builtin_cheri_representable_alignment_mask(stack_bytes) + 1` instead so that everything after `stack_bytes` is at least 16-byte aligned.

dpgao replied:
Well, this is still not quite right. If `stack_bytes` is sufficiently small (say 16) then the padding would be 1 and `struct call_frame` would still be misaligned. We should probably write `__align_up(~__builtin_cheri_representable_alignment_mask(stack_bytes), _Alignof(struct call_frame))` instead.

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -740,6 +755,12 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
    * points just after the alloca with the curretn stack bounds.
    */
   stack = __builtin_cheri_address_set(__builtin_cheri_stack_get(), (ptraddr_t)stack);
```
This actually seems pretty dodgy. Yes, when we call the closure we want CSP's bounds to cover the entire stack, but we don't want to be using an unbounded capability in here, since we really should be fitting everything inside the alloca...

dpgao commented on src/aarch64/ffi.c:
```diff
@@ -740,6 +755,12 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+  stack = __align_up(stack,
+    ~__builtin_cheri_representable_alignment_mask(stack_bytes) + 1);
 #endif
   frame = (void*)((uintptr_t)stack + (uintptr_t)stack_bytes);
```
This line would also need to be changed to include the padding.

kwitaszczyk commented:
@jrtc27 @dpgao What do you think about the current version? I've fixed the stack padding and added a separate call frame padding. Context, stack and frame capabilities have precise bounds now.

kwitaszczyk commented:
I've met with @dpgao and he made me aware that this patch is simply wrong as it doesn't take into account the alignment of the beginning of the allocation (`context`). We're working on a better patch.

kwitaszczyk commented:
@dpgao and I believe this PR is ready for review now.

We've run libffi's tests and the following tests are currently failing:
```
FAIL: libffi.call/float_va.c -W -Wall -O0 execution test
FAIL: libffi.call/float_va.c -W -Wall -O2 execution test
FAIL: libffi.call/va_2.c -W -Wall -O0 execution test
FAIL: libffi.call/va_2.c -W -Wall -O2 execution test
FAIL: libffi.call/va_3.c -W -Wall -O0 execution test
FAIL: libffi.call/va_3.c -W -Wall -O2 execution test
FAIL: libffi.call/va_struct1.c -W -Wall -O0 execution test
FAIL: libffi.call/va_struct1.c -W -Wall -O2 execution test
FAIL: libffi.call/va_struct2.c -W -Wall -O0 execution test
FAIL: libffi.call/va_struct2.c -W -Wall -O2 execution test
FAIL: libffi.call/va_struct3.c -W -Wall -O0 execution test
FAIL: libffi.call/va_struct3.c -W -Wall -O2 execution test
```
These tests were failing on 23.11 as well. We've noticed that libffi wasn't adapted to always use c9 for variadic arguments regardless of the number of arguments. The pure-capability ABI was changed to use c9 for varargs with less than 9 arguments before c18n-related changes in CHERI LLVM for Morello. We'd like to fix these test cases independently of this PR.

jrtc27 commented:
Yeah that plan’s fine. I’ll try and take a look tonight.

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -728,34 +737,95 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+   *
+   * On CHERI-extended architectures, the allocation uses gaps between
+   * sub-objects that have the size equal to the alignment of a sub-object minus
+   * one. The one byte optimisation can be * introduced as for each sub-object
```
* in the middle of the line

jrtc27 replied:
Also, technically you can do better than this by using the known minimum alignment for the end of the previous subobject. But also all this really needs is normal object layout computation where the the members are padded for representability and then overaligned for representability, and it would probably be easier to understand and less messy if you pretend to be a compiler doing struct layout. Something like:
```
size_t length = 0;

size_t call_context_length = __builtin_cheri_round_representable_length(sizeof(struct call_context));
length = __align_up(length, _Alignof(struct call_context));
length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(call_context_length) + 1);
size_t call_context_offset = length;
length += call_context_length;

size_t stack_bytes_length = __builtin_cheri_round_representable_length(stack_bytes);
length = __align_up(length, _Alignof(max_align_t));
length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(stack_bytes_length) + 1);
size_t stack_bytes_offset = length;
length += stack_bytes_length;

size_t call_frame_length = __builtin_cheri_round_representable_length(sizeof(struct call_frame));
length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(call_frame_length) + 1);
size_t call_frame_offset = length;
length += call_frame_length;

size_t rsize_length = __builtin_cheri_round_representable_length(rsize);
length = __align_up(length, _Alignof(max_align_t));
length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(rsize_length) + 1);
size_t rsize_offset = length;
length += rsize_length;
```

Then you just `alloca(length)`.

Or with a macro for the repetition:

```
#define LAYOUT_SUBOBJECT(name, size, align) \
    size_t name##_length = __builtin_cheri_round_representable_length(size); \
    length = __align_up(length, align); \
    length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(name##_length) + 1); \
    size_t name##_offset = length; \
    length += name##_length

LAYOUT_SUBOBJECT(call_context, sizeof(struct call_context), _Alignof(struct call_context));
LAYOUT_SUBOBJECT(stack_bytes, stack_bytes, _Alignof(max_align_t));
LAYOUT_SUBOBJECT(call_frame, sizeof(struct call_frame), _Alignof(struct call_frame));
LAYOUT_SUBOBJECT(rsize, rsize, _Alignof(max_align_t));
```

A lot of that should be able to be constant folded, although Morello LLVM is missing a bunch of that logic in its backend so perhaps not. If you indirect the builtins via stubs that are the identity and constant 1 respectively that code also works just fine on non-CHERI so you don't need to #ifdef it. You may want to adapt the macro to take the actual variable names and declare them at the start of the function though:

```
#define LAYOUT_SUBOBJECT(subobj_length, subobj_offset, size, align) \
    subobj_length = __builtin_cheri_round_representable_length(size); \
    length = __align_up(length, align); \
    length = __align_up(length, ~__builtin_cheri_representable_alignment_mask(subobj_length) + 1); \
    subobj_offset = length; \
    length += subobj_length

LAYOUT_SUBOBJECT(call_context_length, call_context_offset, sizeof(struct call_context), _Alignof(struct call_context));
LAYOUT_SUBOBJECT(stack_bytes_length, stack_bytes_offset, stack_bytes, _Alignof(max_align_t));
LAYOUT_SUBOBJECT(call_frame_length, call_frame_offset, sizeof(struct call_frame), _Alignof(struct call_frame));
LAYOUT_SUBOBJECT(rsize_length, rsize_offset, rsize, _Alignof(max_align_t));
```

Part of the benefit of this is that you only do the representability rounding in one place, rather than the current approach where you compute it once to figure out how much space to allocate and then re-round things as you go with the resulting pointer, hoping that the calculations match up. By fully computing the layout up-front and reusing the same values from the calculation you know it'll match up, so long as the original caluclation was correct, which is easy to verify, especially with the macro.

dpgao replied:
The reason we re-round things as we go is because we worked on the assumption that the pointer returned by `alloca` might be as little as 16-byte aligned, which, on reflection, is clearly too weak.

jrtc27 replied:
It'll be as aligned as the entire allocation needs to be representable, which will be at least as aligned as any sub-object needs for representability

kwitaszczyk replied:
Thank you! This is way cleaner. I've changed it a bit to call `_Alignof()` in the macro, and I've introduced additional macros not to ifdef code as you suggested.

I've added you as a co-author of this change.

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -728,34 +737,95 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+   * All capabilities are derived from context at this point. Set the bounds
+   * now.
+   */
+  context = __builtin_cheri_bounds_set_exact(context, sizeof(*context));
```
You can do this immediately after `cursor = (uintptr_t)context + sizeof(*context);`?

kwitaszczyk replied:
I wanted to separate this from handling particular sub-objects, but I guess the current version using macros is much cleaner.

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -700,6 +705,10 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
   size_t stack_bytes, rtype_size, rsize;
+#ifdef __CHERI_PURE_CAPABILITY__
+  size_t frame_padding, rvalue_padding, stack_padding;
+  uintptr_t cursor;
```
Use `char *` not `uintptr_t` for pointer arithmetic

kwitaszczyk replied:
I've applied this suggestion. Could you tell why it's preferred over `uintptr_t`? It seemed more natural to me to use a data type that is not specific to a data type of referenced data (like `char` in this case).

jrtc27 replied:
My basic philosophy is "don't use uintptr_t unless you need its added functionality". It's mostly about using the right type for the job; uintptr_t is a big hammer that you don't always need, and can have compiler optimisation implications, though that's not hugely important here. Note that char is special in C and can be used to alias any object.

kwitaszczyk replied:
Thanks for clarifying!

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -927,6 +997,29 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+   * pointers to the objects in the on-stack arguments.
+   *
+   * We can use precise bounds for the stack capability because it is already
+   * aligned to a larger size than its effective bounds.
```
But nsaa technically could be an unrepresentable length in absurd extreme cases even if the base is sufficiently aligned.

kwitaszczyk replied:
Could you tell what would be an example of that, and if the assertions that are below this comment don't help with that?

jrtc27 replied:
My point is you can round up the length to get a representable subset of the allocation, but the length itself might not be representable without rounding. To use smaller numbers, say state.nsaa is 1000 because of how many arguments you had, but that needs rounding up to 1016 to be representable. The assertions mean that there is a representable prefix of the stack_length-long region, but you can't bounds_set_exact here in the case that state.nsaa needs rounding up to be representable (which will still be <= stack_length).

kwitaszczyk replied:
Oh, right. Thanks for clarifying this. I hope it looks better now.

jrtc27 replied:
Yes, that looks fine now, thanks

kwitaszczyk commented:
I've added a separate commit that applies current suggestions so that you can see what actually changed. I will fixup it into a single commit before merging this PR.

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -728,35 +752,51 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
-  stack = __builtin_cheri_address_set(__builtin_cheri_stack_get(), (ptraddr_t)stack);
+  context = alloca (total_length);
+
+  cursor = (char *)(uintptr_t)context;
```
Don't need the uintptr_t cast

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -728,35 +752,51 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+  context = alloca (total_length);
+
+  cursor = (char *)(uintptr_t)context;
+  context = (uintptr_t)(cursor + context_offset);
```
Don't need the uintptr_t cast for these, though should cast to the right type

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -33,6 +33,8 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 #endif
 #include <tramp.h>
 
+#define max(a,b) ((a) > (b) ? (a) : (b))
```
Not used any more

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -728,35 +750,51 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
-  stack = __builtin_cheri_address_set(__builtin_cheri_stack_get(), (ptraddr_t)stack);
+  total_length = 0;
+  SUBOBJECT_LAYOUT(total_length, context_length, context_offset,
+    sizeof(*context), struct call_context);
```
typeof(*context) then? It's C23, but has existed for a long time as a GNU extension.

kwitaszczyk replied:
It seems to work well with llvm-morello. Thanks!

jrtc27 commented on src/aarch64/ffi.c:
```diff
@@ -927,10 +965,37 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+    FFI_ASSERT(stack == __align_up(stack, SUBOBJECT_ALIGNMENT(state.nsaa)));
+    context->x9 = (XREG)__builtin_cheri_bounds_set_exact(stack,
+      SUBOBJECT_LENGTH(state.nsaa));
+  } else if (isvariadic) {
```
Might as well always initialise it rather than leave it as 0xaa.... from the padding, given we unconditionally use it in ffi_call_SYSV?

kwitaszczyk replied:
Sounds good.

jrtc27 reviewed (approved):
Subject to squashing and the x9 comment

kwitaszczyk commented:
> Subject to squashing and the x9 comment

Thanks.

@dpgao Any further comments?

dpgao commented on src/aarch64/ffi.c:
```diff
@@ -699,10 +702,29 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+#define SUBOBJECT_LENGTH(size)						\
+	__builtin_cheri_round_representable_length(size)
+#else
+#define SUBOBJECT_ALIGNMENT 1
```
I think it needs to take an argument here, otherwise it won't pre-process?
```suggestion
#define SUBOBJECT_ALIGNMENT(length) 1
```

jrtc27 replied:
Oops, yes. @kwitaszczyk while you're changing this, could you rename SUBOBJECT_LENGTH's parameter to be consistent between the two? I wasn't going to nitpick it as the only change, but now it's not :)

kwitaszczyk replied:
> I think it needs to take an argument here, otherwise it won't pre-process?

Yes, I didn't run an aarch64 build. Thanks!

kwitaszczyk replied:
> Oops, yes. @kwitaszczyk while you're changing this, could you rename SUBOBJECT_LENGTH's parameter to be consistent between the two? I wasn't going to nitpick it as the only change, but now it's not :)

Just to make sure: do you mean that `SUBOBJECT_LENGTH()` and `SUBOBJECT_ALIGNMENT()` should use the same argument name, or they should use different argument name but the same across ABIs they're compiled for (regardless of `__CHERI_PURE_CAPABILITY__`)? It was broken in both cases. :) I'd be leaning to keep the latter as `SUBOBJECT_LAYOUT()` passes a requested size to `SUBOBJECT_LENGTH()` but it passes a representable length to `SUBOBJECT_ALIGNMENT()`.

jrtc27 replied:
Specifically the latter, though not doing the former as well does look a bit odd, even if you want to make a distinction between size and length (which I don't think exists, because a size is a length, even if not representable).

kwitaszczyk replied:
Sounds good. It's fixed now.

dpgao reviewed (approved):
All looks good to me except that one macro.

kwitaszczyk commented:
I have re-run builds and tests for all ABIs and they look correct now (except for aarch64c{,b} tests that we decided not to work on within this PR).

arichardson commented on src/aarch64/ffi.c:
```diff
@@ -699,10 +702,29 @@ ffi_call_int (ffi_cif *cif, void (*fn)(void), void *orig_rvalue,
+#endif
+#define SUBOBJECT_LAYOUT(length, subobj_length, subobj_offset, size, type) \
+	subobj_length = SUBOBJECT_LENGTH(size);				\
+	length = __align_up(length, _Alignof(type));			\
```
Doesn't this break the build for non-FreeBSD? __align_up is not defined there.

jrtc27 replied:
I guess so, though given where this is being used that’s not really a concern. But there seems to be an FFI_ALIGN if we want to support non-FreeBSD.

kwitaszczyk replied:
I've added https://github.com/CTSRD-CHERI/libffi/pull/3.

Unfortunately, I don't have any Linux/arm64 host to check if there any other issues there. https://github.com/CTSRD-CHERI/libffi/pull/3 doesn't introduce any regressions on CheriBSD/Morello.

kwitaszczyk replied:
I'll try to get a separate hard drive for my Morello box and install Morello Linux to test this, but that can take a while.
